lua/null-ls/helpers/cache.lua

```diff
@@ -25,6 +25,27 @@ M.by_bufnr = function(cb)
     end
 end
 
+--- creates a function that caches the output of a callback, indexed by project root
+---@param cb function
+---@return fun(params: NullLsParams): any
+M.by_bufroot = function(cb)
+    -- assign next available key, since we just want to avoid collisions
+    local key = next_key
+    M.cache[key] = {}
+    next_key = next_key + 1
+
+    return function(params)
+        local root = params.root
+        -- if we haven't cached a value yet, get it from cb
+        if M.cache[key][root] == nil then
+            -- make sure we always store a value so we know we've already called cb
+            M.cache[key][root] = cb(params) or false
+        end
+
+        return M.cache[key][root]
```

Comment:
Maybe here you could also have a temporary directory for all the gcroots. You can use the store path as a symlink name. Than you only need to figure out how to clean it when the plugins gets unloaded/shutdown.

Reply:
I sure would love to avoid managing a temp directory and the associated stress of cleaning it up.

How do you feel about reworking the caching logic to be smart enough to rebuild (see my comment above)?

Reply:
We're now storing things in a tempdir. I didn't put it here, because this `by_bufroot` function is supposed to be a generic helper that's unaware of `nix`.